content/blog/learn-pulumi-inputs-outputs/index.md

```diff
@@ -0,0 +1,160 @@
+        vpc,
+        egress: [{ protocol: "-1", fromPort: 0, toPort: 0, cidrBlocks: [ "0.0.0.0/0" ] }],
+    });
+```
+
+In this part of the code, the application load balancer is created, and port 80 is assigned to the load balancer's listener. Note that we create `publicIps` as an Output typed as a string array, which will hold the virtual machines' public IP addresses. We also call `await` to return the public subnets as an output, because we need that information before creating the virtual machines.
+
+```typescript
+    // Creates an ALB associated with the default VPC for this region and listen on port 80.
+    const alb = new awsx.lb.ApplicationLoadBalancer("web-traffic", { securityGroups: [ sg ] });
+    const listener = alb.createListener("web-listener", { port: 80 });
+    const publicIps: pulumi.Output<string>[] = [];
+    const subnets = await vpc.publicSubnets;
+```
+
+It's important to remember you can't access the values of outputs directly. They might not be known, so for example, printing the value to the console returns an error instead. The most common way of interacting with an output is to pass it to another resource as an input.
+
+In this example,  we've instantiated a virtual machine per subnet, and we're passing the public IP address as its target.
+
+```typescript
+    // For each subnet, and each subnet/zone, create a VM and a listener.
+    for (let i = 0; i < subnets.length; i++) {
+        // Create the instance in the same VPC
+        const vm = new aws.ec2.Instance(`web-${i}`, {
+            ami: aws.getAmi({
+                filters: [
+                    { name: "name", values: [ "ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-*" ] },
+                    { name: "virtualization-type", values: [ "hvm" ] },
+                ],
+                mostRecent: true,
+                owners: [ "099720109477" ], // Canonical
+            }).then(ami => ami.id),
+            instanceType: "m5.large",
+            subnetId: subnets[i].subnet.id,
+            availabilityZone: subnets[i].subnet.availabilityZone,
+            vpcSecurityGroupIds: alb.securityGroups.map(sg => sg.securityGroup.id),
+            userData: `#!/bin/bash
+            echo "Hello World, from Server ${i+1}!" > index.html
+            nohup python -m SimpleHTTPServer 80 &`,
+        }, providerOpts);
+        publicIps.push(vm.publicIp);
+        alb.attachTarget("target-" + i, vm);
+    }
+
+    // Export the resulting URL so that it's easy to access.
+    return { endpoint: listener.endpoint.hostname, publicIPs: publicIps  };
+};
+```
+
+As the diagram earlier depicted, another thing we can easily do with outputs is export them. When we export an output, Pulumi captures that value and makes it easily available on the stack. This includes at the command line, in the pulumi console, and even if you want one stack to depend on another stack's output. You can see it in action with the following script:
```

Comment:
```suggestion
As the diagram earlier depicted, another thing we can easily do with outputs is export them. When we export an output, Pulumi captures that value and makes it easily available on the stack. This includes at the command line, in the Pulumi Console console, and even if you want one stack to depend on another stack's output. You can see it in action with the following script:
```

I don't see that the diagram depicts exported outputs. Maybe the diagram "suggests"? And maybe we could point out _where_ in the code the export is? (Because it's fairly hard to find, actually.)